## **Business Understanding**

Air travel safety is a critical concern for both industry stakeholders and the general public. While aviation is among the safetest modes of transport, accidents still occur.  Aircraft manufacturers, airlines, and regulators must therefore evaluate accident patterns to identify risks and improve safety standards.

**Real World Problem**

A new aviation safety advisory group is seeking data-driven recommendations on which aircraft models and manufacturers demonstrate the lowest safety risks. Their goal is to advise flight schools, private operators, and regulators on selecting aircraft types that minimize accident exposure while maintaining operational efficiency.

**Stakeholders**

- **Flight schools**: need to choose training aircraft that balance affordability and safety for students.  
- **Private aircraft owners**: want to make informed decisions when purchasing or operating aircraft.  
- **Regulatory bodies**: require evidence-based insights to guide safety recommendations and certification policies.  
- **Insurance companies**: benefit from understanding which aircraft types carry lower or higher risk.  

**Business Value** 
 
By analyzing accident data, this project will help stakeholders identify:  
1. Which aircraft models and manufacturers have lower accident rates.  
2. How accident severity differs across aircraft types.  
3. What operational factors (e.g., time of day, weather, number of engines) contribute to accident risk.  

**Guiding Business Questions**  

1. Which aircraft manufacturers and models have the lowest accident rates in the last 10–20 years?  
2. How severe are accidents associated with different aircraft types (fatal vs. non-fatal)?  
3. What actionable recommendations can be made about which aircraft types are best suited for minimizing safety risks?  

The findings will directly inform safety-focused decision-making in the aviation sector, with practical implications for reducing accident risks and improving long-term trust in air travel.

## **Data Understanding**

 We explore the dataset to understand its structure, key variables, and potential issues (missing values, duplicates). This helps us assess whether the dataset is suitable for answering our business questions about aviation safety.

This data set provides a record of aircraft accidents that occurred between **1919 - 2023**.


The dataset contains **23,967 rows** and **9 columns**. Each row represents an aviation accident/incident report. The columns provide details about the accident. 

**Columns Overview**
- Date – when the accident/incident occurred
- Location – where it occurred
- Country – country of occurrence
- Type – aircraft type involved
- Registration – aircraft registration number
- Operator – organization/airline operating the aircraft
- Aboard/Fatalities – number of people aboard and number of fatalities
- Cate (Accident Category) – classification of the occurrence
- Year – extracted year of occurrencE

The accident categories are :
- A - Accident
- I - Incident
- H - Hijacking
- C - Criminal occurrence
- O - Other occurrence
- U - type of occurence unknown
- 1 - hull-loss
- 2 - repairable damage

**Initial Observations**

- The dataset spans over a century of aviation history (1919–2023), giving a rich view of trends over time.
- The Cate column is essential for distinguishing between accidents, incidents, and other occurrences.
- The Fatalities variable will allow us to evaluate severity and safety across different dimensions (aircraft type, operator, country).
- Some columns (e.g., registration) may not be directly useful for safety insights and could be dropped later.

In [1]:
# Import libraries
import pandas as pd

In [2]:
df_accidents = pd.read_csv('./data/aviation-accident-data-2023-05-16.csv')
# read first 10 rows
df_accidents.head(10)

date             type registration  \
0  date unk.   Antonov An-12B       T-1206   
1  date unk.   Antonov An-12B       T-1204   
2  date unk.   Antonov An-12B       T-1201   
3  date unk.  Antonov An-12BK          NaN   
4  date unk.  Antonov An-12BP   CCCP-11815   
5  date unk.  Antonov An-12BP   CCCP-12172   
6  date unk.     Antonov An-2    CCCP-N574   
7  date unk.     Antonov An-2   CCCP-01216   
8  date unk.   Antonov An-24B     RA-47794   
9  date unk.    Antonov An-26       01 red   

                               operator fatalities             location  \
0                         Indonesian AF        NaN                  NaN   
1                         Indonesian AF        NaN                  NaN   
2                         Indonesian AF        NaN                  NaN   
3                             Soviet AF        NaN  Tiksi Airport (IKS)   
4                             Soviet AF          0  Massawa Airport ...   
5                             Soviet AF        NaN                  NaN   
6  GUSMP, Directorate of Polar Aviation        NaN              unknown   
7                              Aeroflot          0         Chita region   
8                            Russian AF          0                  NaN   
9                             Soviet AF          0    Orenburg Air Base   

           country cat     year  
0  Unknown country  U1  unknown  
1  Unknown country  U1  unknown  
2  Unknown country  U1  unknown  
3           Russia  A1  unknown  
4          Eritrea  A1  unknown  
5           Russia  U1  unknown  
6           Russia  A1  unknown  
7           Russia  A2  unknown  
8           Russia  A1  unknown  
9           Russia  O1  unknown

In [3]:
# shape of dataset
df_accidents.shape

(23967, 9)

In [4]:
# overview of the dataset
df_accidents.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23967 entries, 0 to 23966
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   date          23967 non-null  object
 1   type          23967 non-null  object
 2   registration  22419 non-null  object
 3   operator      23963 non-null  object
 4   fatalities    20029 non-null  object
 5   location      23019 non-null  object
 6   country       23967 non-null  object
 7   cat           23967 non-null  object
 8   year          23967 non-null  object
dtypes: object(9)
memory usage: 1.6+ MB


From the overview of the datasets we see all columns are of the object data type. We also see that some columns like Registration, Operator, Fatalities, and Location have missing values since the non-null count does not add up to the total number of rows **23,967**

## **Data Preparation**

In this section we clean the data by handling missing values, check for duplicates,and removing irrelevant columns.
We also need to change the Fatalities column from an object data type to an integer data type.

In [5]:
# Checking for duplicates
df_accidents.duplicated().value_counts()

False    23852
True       115
dtype: int64

    From the code above we see that the data set has 115 duplicated items that we need to remove.

In [6]:
# remove duplicates and check if the duplicates were removed
df_accidents = df_accidents.drop_duplicates()
df_accidents.duplicated().value_counts()

False    23852
dtype: int64

    The total number of rows has now dropped from **23,967** to **23,852** due to the removal of duplicates.
    We now deal with missing values

In [7]:
df_accidents.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 23852 entries, 0 to 23966
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   date          23852 non-null  object
 1   type          23852 non-null  object
 2   registration  22418 non-null  object
 3   operator      23848 non-null  object
 4   fatalities    20019 non-null  object
 5   location      22920 non-null  object
 6   country       23852 non-null  object
 7   cat           23852 non-null  object
 8   year          23852 non-null  object
dtypes: object(9)
memory usage: 1.8+ MB


    We first drop the registration column since it is not useful at this time.

In [8]:
# Drop the registration column
df_accidents = df_accidents.drop("registration",axis=1)

    There are now a total of seven columns remaining.

In [9]:
df_accidents.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 23852 entries, 0 to 23966
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   date        23852 non-null  object
 1   type        23852 non-null  object
 2   operator    23848 non-null  object
 3   fatalities  20019 non-null  object
 4   location    22920 non-null  object
 5   country     23852 non-null  object
 6   cat         23852 non-null  object
 7   year        23852 non-null  object
dtypes: object(8)
memory usage: 1.6+ MB


    Next we check the number of missing values in the remaining columns.

In [10]:
# Check for missing values
df_accidents.isna().sum()

date             0
type             0
operator         4
fatalities    3833
location       932
country          0
cat              0
year             0
dtype: int64

    The columns with missing values are operator, fatalities and location.
    We start with the operator column.

In [11]:
# Filling in the values in the operator column
# Since the operator column is an object type we check for the unique values to figure out how to fill in the missing values.
df_accidents["operator"].value_counts()

USAAF                                   2604
USAF                                    1118
RAF                                      920
US Navy                                  745
German AF                                411
                                        ... 
Bohlke                                     1
Táxi Aéreo Pinhal                        1
Country Flyin                              1
AZZA Transport, op.for Sudan Airways       1
Gogal Air Service                          1
Name: operator, Length: 6017, dtype: int64

    We will now fill in the missing values with mode of the column.

In [15]:
operator_mode = df_accidents["operator"].mode()[0]
operator_mode

'USAAF'

In [17]:
df_accidents["operator"] = df_accidents["operator"].fillna(operator_mode)

    We now check if the operator column has any missing values.

In [18]:
df_accidents.isna().sum()

date             0
type             0
operator         0
fatalities    3833
location       932
country          0
cat              0
year             0
dtype: int64

     Next we tackle the location column

In [19]:
# We will still use the mode function for this cause the column is an object type column
df_accidents["location"].value_counts()

unknown                272
Havana-José Mar...    105
Miami Internatio...     71
Beirut Internati...     48
Unknown                 44
                      ... 
between Youngsto...      1
Fazenda Matary, MA       1
Timisoara Airpor...      1
near Tehachapi, CA       1
Alken                    1
Name: location, Length: 14608, dtype: int64

In [20]:
df_accidents["location"] = df_accidents["location"].fillna(df_accidents["location"].mode()[0])

In [21]:
df_accidents.isna().sum()

date             0
type             0
operator         0
fatalities    3833
location         0
country          0
cat              0
year             0
dtype: int64

    We now focus on the fatalities column changing it from object type to integer type and filling in the missing values.

In [23]:
# Checking for unique values
print(df_accidents["fatalities"].unique()[:50])

[nan '0' '14' '1' '4' '2' '5' '6' '3' '7' '5+ 1' '8' '16' '12' '9' '15'
 '14+ 1' '10' '11' '35+ 9' '13' '2+ 2' '17' '19' '20' '0+ 1' '18' '13+ 1'
 '25' '29' '23' '5+ 7' '24' '36' '26' '22' '27' '12+ 1' '21' '7+ 6' '35'
 '32' '31' '2+ 1' '34' '28' '20+ 5' '0+ 2' '30' '41']


In [24]:
# Replacing nan with '0'
df_accidents['fatalities'].fillna(0, inplace=True)

In [25]:
print(df_accidents["fatalities"].unique()[:50])

[0 '0' '14' '1' '4' '2' '5' '6' '3' '7' '5+ 1' '8' '16' '12' '9' '15'
 '14+ 1' '10' '11' '35+ 9' '13' '2+ 2' '17' '19' '20' '0+ 1' '18' '13+ 1'
 '25' '29' '23' '5+ 7' '24' '36' '26' '22' '27' '12+ 1' '21' '7+ 6' '35'
 '32' '31' '2+ 1' '34' '28' '20+ 5' '0+ 2' '30' '41']


In [ ]:
# Changing the column from object to integer and fill in the missing values
df_accidents['fatalities'] = df_accidents['fatalities'].apply(lambda x: sum(map(int, x.replace(' ', '').split('+'))) if '+' in str(x) else int(x))


In [27]:
df_accidents.isna().sum()

date          0
type          0
operator      0
fatalities    0
location      0
country       0
cat           0
year          0
dtype: int64

In [28]:
df_accidents.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 23852 entries, 0 to 23966
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   date        23852 non-null  object
 1   type        23852 non-null  object
 2   operator    23852 non-null  object
 3   fatalities  23852 non-null  int64 
 4   location    23852 non-null  object
 5   country     23852 non-null  object
 6   cat         23852 non-null  object
 7   year        23852 non-null  object
dtypes: int64(1), object(7)
memory usage: 1.6+ MB


**To explain the code:**

- df_accidents['fatalities'] = ...Assigns the result back into the fatalities column.
- .apply(lambda x: ... )- Runs the anonymous function (lambda) once on each cell value x of the column.
- if '+' in str(x)- Converts x to a string and checks if it contains the '+' character.
- x.replace(' ', '')- Removes spaces from the string so splitting is clean.
- .split('+') - Splits the string on '+' into parts ("5+1" to ['5', '1']).
- map(int, ...)- Converts each split substring to an integer.
- sum(...)- Adds the integers together.
- else int(x)- If there is no '+' sign, the code converts the value directly to an integer ( "14" to 14).